File: expo-maps/ios/ExpoMaps/GoogleMaps/GoogleMapsView.swift
import GoogleMaps
import GooglePlaces
import ExpoModulesCore

public final class GoogleMapsView: UIView, ExpoMapView {

  private let mapView: GMSMapView
  private let googleMapsViewDelegate: GoogleMapsViewDelegate
  private let controls: GoogleMapsControls
  private let markers: GoogleMapsMarkers
  private let clusters: GoogleMapsClusters
  private let googleMapsClusterManagerDelegate: GoogleMapsClusterManagerDelegate
  private let gestures: GoogleMapsGestures
  private let polygons: GoogleMapsPolygons
  private let polylines: GoogleMapsPolylines
  private let circles: GoogleMapsCircles
  private let kmls: GoogleMapsKMLs
  private let geojsons: GoogleMapsGeoJsons
  private let overlays: GoogleMapsOverlays
  private let places: GoogleMapsPlaces
  private var wasInitialCameraPositionSet = false
  private let heatmaps: GoogleMapsHeatmaps
  public var clickablePOIs = true
  private let googleMapsMarkersManager: GoogleMapsMarkersManager = GoogleMapsMarkersManager()

  // TODO: change to proper types from "[String: Any?]" when conversion from records gets implemented for iOS
  @Event var onMapLoaded: Callback<String?>
  @Event var onMapPress: Callback<[String: Any?]>
  @Event var onDoublePress: Callback<[String: Any?]>
  @Event var onLongPress: Callback<[String: Any?]>
  @Event var onRegionChange: Callback<[String: Any?]>
  @Event var onRegionChangeStarted: Callback<[String: Any?]>
  @Event var onRegionChangeComplete: Callback<[String: Any?]>
  @Event var onPoiClick: Callback<[String: Any?]>
  @Event var onMarkerPress: Callback<[String: Any?]>
  @Event var onMarkerDrag: Callback<[String: Any?]>
  @Event var onMarkerDragStarted: Callback<[String: Any?]>
  @Event var onMarkerDragComplete: Callback<[String: Any?]>
  @Event var onClusterPress: Callback<[String: Any?]>
  @Event var onLocationButtonPress: Callback<[String: Any?]>
  @Event var onLocationDotPress: Callback<[String: Any?]>
  @Event var onLocationChange: Callback<[String: Any?]>

  init(sendEvent: @escaping (String, [String: Any?]) -> Void) {
    // just for now we do authentication here
    // should be moved to module's function
    GoogleMapsView.initializeGoogleMapsServices()

    // random initial camera position
    // TODO: use prop as a source for initial camera position
    let camera = GMSCameraPosition.camera(withLatitude: 51.5, longitude: 0, zoom: 4.0)
    mapView = GMSMapView.map(withFrame: CGRect.zero, camera: camera)
    googleMapsViewDelegate = GoogleMapsViewDelegate(sendEvent: sendEvent, googleMapsMarkersManager: googleMapsMarkersManager)
    mapView.delegate = googleMapsViewDelegate
    mapView.autoresizingMask = [.flexibleWidth, .flexibleHeight]
    controls = GoogleMapsControls(mapView: mapView)
    markers = GoogleMapsMarkers(mapView: mapView, googleMapsMarkersManager: googleMapsMarkersManager)
    googleMapsClusterManagerDelegate = GoogleMapsClusterManagerDelegate(googleMapsMarkersManager: googleMapsMarkersManager)
    clusters = GoogleMapsClusters(mapView: mapView, googleMapsMarkersManager: googleMapsMarkersManager, googleMapsClusterManagerDelegate: googleMapsClusterManagerDelegate, googleMapsViewDelegate: googleMapsViewDelegate)
    gestures = GoogleMapsGestures(mapView: mapView)
    polygons = GoogleMapsPolygons(mapView: mapView)
    polylines = GoogleMapsPolylines(mapView: mapView)
    circles = GoogleMapsCircles(mapView: mapView)
    kmls = GoogleMapsKMLs(mapView: mapView)
    geojsons = GoogleMapsGeoJsons(mapView: mapView)
    overlays = GoogleMapsOverlays(mapView: mapView)
    heatmaps = GoogleMapsHeatmaps(mapView: mapView)
    places = GoogleMapsPlaces(mapView: mapView, markers: markers)

    super.init(frame: CGRect.zero)
    googleMapsViewDelegate.expoMapView = self
    googleMapsClusterManagerDelegate.setOnClusterPress(onClusterPress: onClusterPress)
    mapView.addObserver(googleMapsViewDelegate, forKeyPath: "myLocation", context: nil)
    addSubview(mapView)
  }

  required init?(coder: NSCoder) {
    fatalError("init(coder:) has not been implemented")
  }

  // Allows the double tap to work

  public func gestureRecognizer(_ gestureRecognizer: UIGestureRecognizer,
                                _ otherGestureRecognizer: UIGestureRecognizer) -> Bool {
    true
  }

  private static func initializeGoogleMapsServices() {
    guard let path = Bundle.main.path(forResource: "Info", ofType: "plist") else {
      fatalError("Couldn't find file 'Info.plist'.")
    }
    let content = NSDictionary(contentsOfFile: path)
    guard let googleMapsApiKey = content?.object(forKey: "GoogleMapsApiKey") as? String else {
      fatalError("Couldn't find key 'GoogleMapsApiKey' in 'Info.plist'.")
    }
    GMSServices.provideAPIKey(googleMapsApiKey)
    GMSPlacesClient.provideAPIKey(googleMapsApiKey)
  }

  func fetchPlacesSearchCompletions(searchQueryFragment: String, promise: Promise) {
    places.fetchSearchCompletions(searchQueryFragment: searchQueryFragment, promise: promise)
  }

  func createPOISearchRequest(place: String) {
    places.createSearchRequest(place: place)
  }

  func setShowCompass(enable: Bool) {
    controls.setShowCompass(enable: enable)
  }

  func setShowMyLocationButton(enable: Bool) {
    controls.setShowMyLocationButton(enable: enable)
  }

  func setShowLevelPicker(enable: Bool) {
    controls.setShowLevelPicker(enable: enable)
  }

  func setEnabledRotateGestures(enabled: Bool) {
    gestures.setEnabledRotateGesture(enabled: enabled)
  }

  func setEnabledScrollGestures(enabled: Bool) {
    gestures.setEnabledScrollGesture(enabled: enabled)
  }

  func setEnabledTiltGestures(enabled: Bool) {
    gestures.setEnabledTiltGesture(enabled: enabled)
  }

  func setEnabledZoomGestures(enabled: Bool) {
    gestures.setEnabledZoomGesture(enabled: enabled)
  }

  func setMapType(mapType: MapType) {
    var mapViewType: GMSMapViewType
    switch mapType {
    case .hybrid:
      mapViewType = .hybrid
    case .satellite:
      mapViewType = .satellite
    case .terrain:
      mapViewType = .terrain
    case .normal:
      mapViewType = .normal
    }
    mapView.mapType = mapViewType
  }

  func setMapStyle(jsonStyleString: String) {
    if jsonStyleString.count != 0 {
      do {
        mapView.mapStyle = try GMSMapStyle(jsonString: jsonStyleString)
      } catch {
        NSLog("One or more of the map styles failed to load. \(error)")
      }
    } else {
      mapView.mapStyle = nil
    }
  }

  func setMarkers(markerObjects: [MarkerObject]) {
    markers.setMarkers(markerObjects: markerObjects)
  }

  func setPolygons(polygonObjects: [PolygonObject]) {
    polygons.setPolygons(polygonObjects: polygonObjects)
  }

  func setPolylines(polylineObjects: [PolylineObject]) {
    polylines.setPolylines(polylineObjects: polylineObjects)
  }

  func setCircles(circleObjects: [CircleObject]) {
    circles.setCircles(circleObjects: circleObjects)
  }

  func updatePolylines() {
    polylines.updateStrokePatterns()
  }

  func updatePolygons() {
    polygons.updateStrokePatterns()
  }

  func setInitialCameraPosition(initialCameraPosition: CameraPosition) {
    if (!wasInitialCameraPositionSet) {
      let newCameraPosition = GMSCameraPosition(latitude: initialCameraPosition.latitude, longitude: initialCameraPosition.longitude, zoom: Float(initialCameraPosition.zoom))
      if (initialCameraPosition.animate) {
        mapView.animate(to: newCameraPosition)
      } else {
        mapView.camera = newCameraPosition
      }
      wasInitialCameraPositionSet = true
    }
  }

  func setClusters(clusterObjects: [ClusterObject]) {
    clusters.setClusters(clusterObjects: clusterObjects)
  }

  func setEnabledTraffic(enableTraffic: Bool) {
    mapView.isTrafficEnabled = enableTraffic
  }

  func setKMLs(kmlObjects: [KMLObject]) {
    kmls.setKMLs(kmlObjects: kmlObjects)
  }

  func setGeoJsons(geoJsonObjects: [GeoJsonObject]) {
    geojsons.setGeoJsons(geoJsonObjects: geoJsonObjects)
  }

  func setOverlays(overlayObjects: [OverlayObject]) {
    overlays.setOverlays(overlayObjects: overlayObjects)
  }

  func setHeatmaps(heatmapObjects: [HeatmapObject]) {
    heatmaps.setHeatmaps(heatmapObjects: heatmapObjects)
  }

  func setClickablePOIs(clickablePOIs: Bool) {
    self.clickablePOIs = clickablePOIs
  }
}

File: expo-maps/android/src/main/java/expo/modules/maps/googleMaps/GoogleMapsView.kt
package expo.modules.maps.googleMaps

import android.content.Context
import android.widget.LinearLayout
import com.google.android.gms.maps.CameraUpdateFactory
import com.google.android.gms.maps.GoogleMap
import com.google.android.gms.maps.MapView
import com.google.android.gms.maps.OnMapReadyCallback
import com.google.android.gms.maps.model.LatLng
import com.google.android.gms.maps.model.MapStyleOptions
import com.google.maps.android.collections.MarkerManager
import expo.modules.kotlin.Promise
import expo.modules.kotlin.callbacks.callback
import expo.modules.maps.*
import expo.modules.maps.googleMaps.events.GoogleMapsEventEmitterManager
import expo.modules.maps.interfaces.ExpoMapView
import expo.modules.maps.records.*
import kotlinx.coroutines.*
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.collectLatest

class GoogleMapsView(context: Context) : LinearLayout(context), OnMapReadyCallback, ExpoMapView {

  private val mapView: MapView = MapView(context)
  private lateinit var googleMap: GoogleMap
  private lateinit var controls: GoogleMapsControls
  private lateinit var gestures: GoogleMapsGestures
  private lateinit var markers: GoogleMapsMarkers
  private lateinit var clusters: GoogleMapsClusters
  private lateinit var polygons: GoogleMapsPolygons
  private lateinit var polylines: GoogleMapsPolylines
  private lateinit var circles: GoogleMapsCircles
  private lateinit var kmls: GoogleMapsKMLs
  private lateinit var geojsons: GoogleMapsGeoJsons
  private lateinit var markerManager: MarkerManager
  private lateinit var overlays: GoogleMapsOverlays
  private lateinit var heatmaps: GoogleMapsHeatmaps
  private lateinit var places: GoogleMapsPlaces
  private lateinit var callbacks: GoogleMapsCallbacks

  private val mapReady = MutableStateFlow(false)
  private var wasInitialCameraPositionSet = false

  private val onMapLoaded by callback<Unit>()
  private val onMapClick by callback<LatLngRecord>()
  private val onLongPress by callback<LatLngRecord>()
  private val onRegionChange by callback<CameraPositionRecord>()
  private val onRegionChangeStarted by callback<CameraPositionRecord>()
  private val onRegionChangeComplete by callback<CameraPositionRecord>()
  private val onPoiClick by callback<PointOfInterestRecord>()
  private val onMarkerPress by callback<MarkerObject>()
  private val onMarkerDrag by callback<MarkerObject>()
  private val onMarkerDragStarted by callback<MarkerObject>()
  private val onMarkerDragComplete by callback<MarkerObject>()
  private val onClusterPress by callback<ClusterRecord>()


  val lifecycleEventListener = MapViewLifecycleEventListener(mapView)

  init {
    mapView.onCreate(null)
    mapView.getMapAsync(this)
    mapView.onStart()
    mapView.onResume()
    addView(mapView)
  }

  override fun onMapReady(googleMap: GoogleMap) {
    this.googleMap = googleMap
    markerManager = MarkerManager(googleMap)
    controls = GoogleMapsControls(googleMap)
    gestures = GoogleMapsGestures(googleMap)
    markers = GoogleMapsMarkers(googleMap, markerManager)
    clusters = GoogleMapsClusters(context, googleMap, markerManager, onClusterPress, onMarkerPress)
    polygons = GoogleMapsPolygons(googleMap)
    polylines = GoogleMapsPolylines(googleMap)
    circles = GoogleMapsCircles(googleMap)
    kmls = GoogleMapsKMLs(context, googleMap)
    geojsons = GoogleMapsGeoJsons(googleMap)
    overlays = GoogleMapsOverlays(googleMap)
    heatmaps = GoogleMapsHeatmaps(googleMap)
    places = GoogleMapsPlaces(context, googleMap, markers)
    callbacks = GoogleMapsCallbacks(googleMap)

    CoroutineScope(Dispatchers.Default).launch {
      mapReady.emit(true)
    }

    setupCallbacks()
  }

  fun setShowZoomControl(enable: Boolean) {
    updateMap {
      controls.setShowZoomControl(enable)
    }
  }

  fun setShowCompass(enable: Boolean) {
    updateMap {
      controls.setShowCompass(enable)
    }
  }

  fun setShowMapToolbar(enable: Boolean) {
    updateMap {
      controls.setShowMapToolbar(enable)
    }
  }

  fun setShowMyLocationButton(enable: Boolean) {
    updateMap {
      controls.setShowMyLocationButton(enable)
    }
  }

  fun setShowLevelPicker(enable: Boolean) {
    updateMap {
      controls.setShowLevelPicker(enable)
    }
  }

  fun setEnabledRotateGestures(enabled: Boolean) {
    updateMap {
      gestures.setEnabledRotateGesture(enabled)
    }
  }

  fun setEnabledScrollGestures(enabled: Boolean) {
    updateMap {
      gestures.setEnabledScrollGesture(enabled)
    }
  }

  fun setEnabledTiltGestures(enabled: Boolean) {
    updateMap {
      gestures.setEnabledTiltGesture(enabled)
    }
  }

  fun setEnabledZoomGestures(enabled: Boolean) {
    updateMap {
      gestures.setEnabledZoomGesture(enabled)
    }
  }

  fun fetchPlacesSearchCompletions(searchQueryFragment: String, promise: Promise) {
    updateMap {
      places.fetchSearchCompletions(searchQueryFragment, promise)
    }
  }

  fun createPlaceSearchRequest(place: String) {
    updateMap {
      places.createSearchRequest(place)
    }
  }

  fun setClickablePOIs(arePOIClickable: Boolean) {
    updateMap {
      places.setClickablePOIs(arePOIClickable)
    }
  }

  override fun setMapType(mapType: MapType) {
    val googleMapType = when (mapType) {
      MapType.Normal -> GoogleMap.MAP_TYPE_NORMAL
      MapType.Terrain -> GoogleMap.MAP_TYPE_TERRAIN
      MapType.Satellite -> GoogleMap.MAP_TYPE_SATELLITE
      MapType.Hybrid -> GoogleMap.MAP_TYPE_HYBRID
    }

    updateMap {
      googleMap.mapType = googleMapType
    }
  }

  fun setMapStyle(jsonStyleString: String) {
    if (jsonStyleString.isNotEmpty()) {
      updateMap {
        googleMap.setMapStyle(MapStyleOptions(jsonStyleString))
      }
    } else {
      updateMap {
        googleMap.setMapStyle(null)
      }
    }
  }

  override fun setMarkers(markerObjects: Array<MarkerObject>) {
    updateMap {
      markers.setMarkers(markerObjects)
    }
  }

  override fun setPolygons(polygonObjects: Array<PolygonObject>) {
    updateMap {
      polygons.setPolygons(polygonObjects)
    }
  }

  override fun setPolylines(polylineObjects: Array<PolylineObject>) {
    updateMap {
      polylines.setPolylines(polylineObjects)
    }
  }

  override fun setCircles(circleObjects: Array<CircleObject>) {
    updateMap {
      circles.setCircles(circleObjects)
    }
  }


  override fun setInitialCameraPosition(initialCameraPosition: CameraPosition) {
    if (!wasInitialCameraPositionSet) {
      updateMap {
        val cameraUpdate = CameraUpdateFactory.newLatLngZoom(
          LatLng(initialCameraPosition.latitude, initialCameraPosition.longitude),
          initialCameraPosition.zoom.toFloat()
        )
        if (initialCameraPosition.animate) {
          googleMap.animateCamera(cameraUpdate)
        } else {
          googleMap.moveCamera(cameraUpdate)
        }
      }
      wasInitialCameraPositionSet = true
    }
  }

  override fun setClusters(clusterObjects: Array<ClusterObject>) {
    updateMap {
      clusters.setClusters(clusterObjects)
    }
  }

  override fun setEnabledTraffic(enableTraffic: Boolean) {
    updateMap {
      googleMap.isTrafficEnabled = enableTraffic
    }
  }

  override fun setKMLs(kmlObjects: Array<KMLObject>) {
    updateMap {
      kmls.setKMLs(kmlObjects)
    }
  }

  override fun setGeoJsons(geoJsonObjects: Array<GeoJsonObject>) {
    updateMap {
      geojsons.setGeoJsons(geoJsonObjects)
    }
  }

  override fun setHeatmaps(heatmapObjects: Array<HeatmapObject>) {
    updateMap {
      heatmaps.setHeatmaps(heatmapObjects)
    }
  }

  override fun setOverlays(overlayObjects: Array<OverlayObject>) {
    updateMap {
      overlays.setOverlays(overlayObjects)
    }
  }

  fun registerEvents(mapsEventEmitterManager: GoogleMapsEventEmitterManager) {
    updateMap {
      mapsEventEmitterManager.createEmitters(googleMap)
      clusters.googleMapsEventEmitterManager = mapsEventEmitterManager
      clusters.setOnCameraIdleListener(mapsEventEmitterManager.mapsEventEmitterCameraMoveEnded)
    }
  }

  private fun setupCallbacks() {
    callbacks.setupOnMapClick(onMapClick)
    callbacks.setupOnMapLoaded(onMapLoaded)
    callbacks.setupOnRegionChange(onRegionChange)
    callbacks.setupOnRegionChangeStarted(onRegionChangeStarted)
    callbacks.setupOnRegionChangeComplete(onRegionChangeComplete)
    callbacks.setupOnPoiClick(onPoiClick)
    callbacks.setupOnLongPress(onLongPress)

    markers.setOnMarkerPressListener(onMarkerPress)
    markers.setOnMarkerDragListeners(onMarkerDrag, onMarkerDragStarted, onMarkerDragComplete)

  }

  /*
      Calls function provided as an argument when OnMapReadyCallback fires,
      subscribes to StateFlow in a background but calls lambda on a main thread.
      After calling lambda the subscription is canceled.
      StateFlow holds the latest value so even if updateMap is called after
      OnMapReadyCallback, StateFlow emits the latest value letting provided lambda to be executed.
  */
  private fun updateMap(update: () -> Unit) {
    CoroutineScope(Dispatchers.IO).launch {
      mapReady.collectLatest {
        if (it) {
          withContext(Dispatchers.Main) {
            update()
          }
          cancel()
        }
      }
    }
  }
}
